### PR TITLE
created signin page

### PR DESCRIPTION
**Pull request by nikitonskii**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,15 +1,19 @@
-import React from 'react';
-import logo from './logo.svg';
-import './App.scss';
-import SignIn from './pages/signIn/SignIn';
+import React from "react";
+
+import logo from "./logo.svg";
+import "./App.scss";
+
+import SignIn from "./pages/signIn/SignIn";
```
Comment (twixDuo) [inline src/App.tsx]: I think better change path. You should create index file in folder ```signIn``` and import there your component ```SignIn``` and ```scss``` and after you can use this path for import ```import SignIn from "./pages/signIn/```

File: src/App.tsx [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
+import SignUp from "./pages/signUp/SignUp";
 
 const App: React.FC = (): JSX.Element => {
   return (
-    <div className='App'>
-      <img src={logo} className='App-logo' alt='logo' />
+    <div className="App">
+      <img src={logo} className="App-logo" alt="logo" />
```
Comment (twixDuo) [inline src/App.tsx]: I think better use name of class ```app-logo```

File: src/App.tsx [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
       <SignIn />
+      <SignUp />
```
Comment (twixDuo) [inline src/App.tsx:14]: how you understand what component you must show ? or in this case you show component ```SignIn``` and ```SignUp``` ?

File: src/pages/signIn/SignIn.tsx [hunk 4]
```diff
@@ -10,10 +10,9 @@ import { config } from "./config";
 
 import { Fields, ValidationFields, ConfigItem } from "./types";
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: I think better create file ```intex.ts``` in folder ```types``` and export there all you types, and after that you can import your type like that ```import * as types from "./types";```

File: src/pages/signIn/config.ts [hunk 5]
```diff
@@ -1,20 +1,4 @@
...
-  {
-    name: 'secondName',
-    type: 'text',
-    label: 'Second name',
-    isRequired: false,
-    placeholder: 'your second name',
-  },
   {
     name: 'email',
     type: 'email',
```
Comment (twixDuo) [inline src/pages/signIn/config.ts]: In this case i think better use  icon ```*``` to check you key ```isRequired``` and dont use this icon with  ``` label: 'Email *',```

File: src/pages/signUp/SignUp.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+import { Fields, ValidationFields, ConfigItem } from "./types";
+
+import * as utils from "../../utils";
+
+const SignUp: React.FC = (): JSX.Element => {
+  const [success, setSuccess] = useState<boolean>(false);
+  const [formFields, setFormFields] = useState<Fields>({});
+  const [isFieldsValid, setIsFieldsValid] = useState<ValidationFields>({});
+
+  useLayoutEffect(() => {
```
Comment (twixDuo) [inline src/pages/signUp/SignUp.tsx]: Can you add short description for this function. what it do  ?

Example: ```// ----- and here your short description ```

File: src/pages/signUp/SignUp.tsx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    const fields: Fields = {};
+    const validationVars: ValidationFields = {};
+
+    config.forEach((field: ConfigItem) => {
+      const { name } = field;
+
+      if (field.isRequired) {
```
Comment (twixDuo) [inline src/pages/signUp/SignUp.tsx]: For optimized your code you can do this.

```
if (field.isRequired) validationVars[name] = false;
fields[name] = "";
```

File: src/pages/signUp/SignUp.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    setFormFields({
+      ...formFields,
+      [name]: value,
+    });
+  };
+
+  const isValidationError = (config: ConfigItem[]): boolean => {
+    const state: ValidationFields = {};
+    let fields = [];
+    config.forEach((field: ConfigItem) => {
```
Comment (twixDuo) [inline src/pages/signUp/SignUp.tsx]: We discussed with you. divide the code into blocks
```
    const state: ValidationFields = {};
    let fields = [];

    config.forEach((field: ConfigItem) => {
```

File: src/pages/signUp/SignUp.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+    setIsFieldsValid(state);
+
+    return [...Object.values(state)].includes(true);
+  };
+
+  const makeFakeRequest = () => {
+    const checkResult = isValidationError(config);
+
+    if (!checkResult) {
```
Comment (twixDuo) [inline src/pages/signUp/SignUp.tsx]: I think better write this code in one line

```
if (!checkResult) setTimeout(() => setSuccess(true), 1000);
```

File: src/App.tsx (hunk 3 above)
Comment (nikitonskii) [reply src/App.tsx:14]: in this case i show both components, iam going to separate it after implementing a routing

File: src/components/Form/Form.tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+}): JSX.Element => {
+  const [formFields, setFormFields] = useState<types.Fields>({});
+  const [isFieldsValid, setIsFieldsValid] = useState<types.ValidationFields>(
+    {},
+  );
+
+  // creates fields dynamicly
+  useLayoutEffect(() => {
+    const fields: types.Fields = {};
+    const validationVars: types.ValidationFields = {};
```
Comment (twixDuo) [inline src/components/Form/Form.tsx]: I think validationFields is better name than validationVars

File: src/components/Form/Form.tsx [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+      if (field === "email") {
+        state[field] = !utils.regExEmail.test(formFields[field]);
+      } else {
+        state[field] = !formFields[field].length;
+      }
+    });
+
+    setIsFieldsValid(state);
+
+    return [...Object.values(state)].includes(true);
```
Comment (twixDuo) [inline src/components/Form/Form.tsx]: I don`t understand  why you use operator spread ?

**Review by twixDuo — CHANGES_REQUESTED** [review]

**Pull request merged**